### ISSUE TITLE
API to allow 3rd-party to create a draft table, and get decklists at the end of the draft

### ISSUE THREAD

**Issue by silasary** [dr4fters/dr4ft#87]
I've been asked about adding limited support to https://github.com/PennyDreadfulMTG/gatherling/, but I would much rather be able to use Dr4ft as the backing for such a feature, rather than re-implementing my own draft simulator.

**Comment by HerveH44** [comment]
Hi,

It seems super interesting. In order to clarify what you would need, I'd like to know what kind of games you would like to create, a draft of the last sets for example? Any sealed, cube, chaos or other mods?

As a quick guess, the parameters for the creation of a draft, we would need at least these informations, for example on a post of our endpoint :
title[string], 
seats[int], 
type[enum:"draft","sealed","cube sealed","cube draft","chaos"], 
sets[stringArray(for example "XLN,XLN,XLN"],
 cube[boolean], isPrivate[boolean], 
fourPack[boolean],
 modernOnly[boolean], 
totalChaos[boolean]

We could then answer with the proper game url, so you could then send it to your users I guess? 

Then to get all the card pools of the users, we don't have that already, but it shouldn't be hard technically to do. The ony thing is that it should be more or less protected, so other players from the draft can't get it. Otherwise any player could know the pool and then the definite cards from any other drafter. I don't know how it could be implemented, maybe with a code sent with the answer for the creation of the game. This code should be then used to retrieve the pools of all drafters. 

In summary, could you be a bit more explicit in what you need and how you would like to have everything implemented? 

Thanks,

Herve

**Comment by silasary** [comment]
How I'd expect it to work:

# POST /api/create
_Intent:  Create a draft room_

Takes in draft parameters, as form-encoded and/or JSON.  For now, let's assume that the consumer knows all the needed values, and doesn't need any special helper stuff.

For my purposes, the settings for "Bots", "Timer", and "Random Seating" are also parameters that should be passed in here.  They should not be configurable by players.

We might also want to pass in a webhook url to ping when draft is completed.

```
title[string],
seats[int],
type[enum:"draft","sealed","cube sealed","cube draft","chaos"],
sets[stringArray(for example "XLN,XLN,XLN"],
cube[boolean],
isPrivate[boolean],
fourPack[boolean],
modernOnly[boolean],
totalChaos[boolean],
bots[boolean],
timer[enum:Fast, Moderate, Slow, Leisurely],
random_seating[bool],
webhook[string],
secret[string] (Optional - autogenerated by dr4ft server?),
```

Returns JSON with
```
{
 "game": <link to game>[string],
 "secret": <random key used for further api calls>[string]
}
```

# GET /api/status?secret=[string]
_Intent: Check the status of the draft, which players are ready, and which are not._

Takes room secret
Returns JSON with 
```
{
 "drafting": boolean,
 "current_pack": int,
 "players": [
  {
   "seat_number": int,
   "player_name": string,
   "ready": boolean,
  }
 ]
}
```

# POST /api/start?secret=[string]
_Intent: Start the draft_

Takes room secret.

# GET /api/deck?secret=[string]&seat=[int|string]
_Intent: Gets decklists at end of draft_
Takes room secret, and either seat number or player name.
Returns JSON with 
```
{
 "seat_number": int,
 "player_name": string,
 "pool": array[string],
}
```

# Other Requirements

The join link should support a playername parameter, so that links provided to each player will automatically be given the appropriate name.

Alternatively, I could use a separate invite link per seat (Although I think this is probably more work)

It might be nice to also support an option to auto-fire the draft when the table becomes full, so that the consumer doesn't need to poll during the pre-draft state.

**Comment by HerveH44** [comment]
Hi @silasary ,

Thanks for your very precise api request. I was inspired by your work to establish a more "restful" api in my opinion which also doesn't break our current configuration.

So this is the api i'm currently working on : 

# Create game
_Intent: Create a draft room_
Method: POST 
Path: /api/games

Takes in draft parameters, as  **JSON**, all of the following parameters are expected in the request body, except the gameId and secret which need to be in the request params.

## Parameters(Body)

| Parameter | Type    | Sample   |
|:-------:|:------:|:-------:|
| title     | string  | myTitle  |
| seats     | int     | 8        |
| type      | string  | "draft","sealed","cube sealed","cube draft","chaos"  |
| cube      | boolean | false    |
| isPrivate | boolean | true     |
| fourPack  | boolean | true     |
| modernOnly| boolean | true     |
| totalChaos| boolean | true     |

_webhook[string]_, not implemented yet

## Returns

Type: **JSON**

| Attribute | Type    | Sample    |      Info       |
|:-------:|:-------:|:--------:|:-----------:| 
| gamePath  | string  | #g/t7fazj | path to game|
| gameId    | string  | t7fazj    | id of game  |
| secret    | string  | a1zf563   | random key used for further api calls |

# Get game status
_Intent: Check the status of the draft, which players are ready, and which are not._
Method: GET 
Path: /api/games/:gameId/status?secret=:secret


## Parameters(Params)
gameId: <id of game>[string]
secret: <secret of the game>[string]

## Returns

Type: **JSON**

| Attribute    | Type    | Sample |  Info           |
|:----------:|:-------:|:-----:|:-----------:| 
| didGameStart | boolean | true   | true if the game started|
| currentPack  | int     | 1      | pack number. 0 when not started  |
| players      | Array   | []     | List of players in the game |

Each player has several attributes

| Attribute     | Type    | Sample   |     Info        |
|:-----------:|:-------:|:-------:|:-----------:| 
|playerName     | string  | dr4fter  | the player name|
|id             | string  |fj0ryrept4| the player Id|
|isReadyToStart | boolean | true     | shows if the player started|
|seatNumber     | int     | 0        | from 0, 0 means the first player|

The playerId is important as two players could share the same name.

### Sample

```javascript
{
    "didGameStart": true,
    "currentPack": 1,
    "players": [
        {
            "playerName": "dr4fter",
            "id": "fj0ryrept4",
            "isReadyToStart": true,
            "seatNumber": 0
        }
    ]
}
```


# Start a Game
_Intent: Start the draft_
Method: POST 
Path: /api/games/:gameId/start?secret=:secret

## Parameters(Params)
gameId: <id of game>[string]
secret: <secret of the game>[string]

## Parameters(Body)

| Parameter     | type    | sample   |
|:-----------:|:-------:|:-------:|
| addBots       | boolean | true     |
| useTimer      | boolean | true     |
| timer         | string  | Fast, Moderate, Slow or Leisurely  |
| shufflePlayers| boolean | false    |

**timer** is mandatory only if **useTimer** is set to _true_

## Returns

Type: **JSON**

Message: "Game _gameId_ successfully started

### Sample
```javascript
{
    "message": "Game pucip3cvv8 successfully started"
}
```

sends a 400 with parameters' errors if some parameters were wrong 

# Get decklist(s)
_Intent: Gets decklists at end of draft_
Method: GET 
Path: /api/games/:gameId/deck?secret=:secret&seat=:seat&id=:playerId

## Parameters(Params)
gameId: <id of game>[string]
secret: <secret of the game>[string]
seat: <seat number>[int]
id: <playerId>[string]

## Returns

If no seat/id has been provided, returns an array with all players of the game

Type: **JSON**

| Attribute  | type   | sample |             |
|:--------:|:------:|:-----:|:-----------:| 
| seatNumber | int    | 0      | seat Number|
| playerName | string | dr4fter| bot or player name  |
| playerId   | string | 1      | player Id |
| pool       | array  | 1      | array of <cards>  |

Each card has several attributes and are classified by pick order.

### Sample
```javascript
[
    {
        "seatNumber": 0,
        "playerName": "bot"
    },
    {
        "seatNumber": 1,
        "playerName": "dr4fter",
        "id": "fj0ryrept4",
        "pool": [
            {
                "color": "white",
                "name": "Imperial Aerosaur",
                "type": "Creature",
                "cmc": 4,
                "text": "Flying\nWhen Imperial Aerosaur enters the battlefield, another target creature you control gets +1/+1 and gains flying until end of turn.",
                "manaCost": "{3}{W}",
                "layout": "normal",
                "isDoubleFaced": false,
                "score": 2.1569606211869106,
                "packSize": 14,
                "code": "XLN",
                "rarity": "uncommon",
                "url": "http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=435165&type=card"
            }
        ]
    }
]
```

**Comment by ZeldaZach** [comment]
@HerveH44 
Start Game: We should return more then successfully started. Include # of bots added and the timer speed maybe?

Get decklist: the "score" param is something internal to dr4ft and our bots, and I'd rather not share that information at this time via an API.

**Comment by HerveH44** [comment]
@ZeldaZach Ok. Now the start Game also shows the number of bots. The pool only show the cards' name and nothing else.

**Comment by silasary** [comment]
Cardname is all I'd need anyway - The rest I'd want to be pulling from my own DB anyway.

API looks good 👍

**Comment by ZeldaZach** [comment]
Sounds good to me then @HerveH44. Returning a list of the cards drafted via name is acceptable. it''s not our job to include card meta data

**Comment by HerveH44** [comment]
Hi @silasary ,

if you have some time, could you review the API too in our deploy base here : https://beta-dr4ft-pr-96.herokuapp.com/ and tell me if you find any problem? We will modify the PR in consequences :) ! 

Thanks, 

Hervé

**Comment by dev-id** [comment]
Thank you for your work on this Hervé.

I did a quick test of the functions on the beta.

The `create` request is not requiring the "sets" or "list" which allows creation of invalid / unstartable games. I'm not sure how to pass the cube list, as `cube` is bool and adding a `list` returns `"cube.list must be a string"` even when it is.

I'm not able to get `deck` to work, using:
`curl --request GET \
  --url 'https://beta-dr4ft-pr-96.herokuapp.com/api/games/xkd5lniht2q/deck?seat=1&secret=052ea16d-cf54-4a1a-a9fe-a7b5acf44b7d'`

Here's the status for the same draft:
```
curl --request GET \
  --url 'https://beta-dr4ft-pr-96.herokuapp.com/api/games/xkd5lniht2q/status?secret=%22052ea16d-cf54-4a1a-a9fe-a7b5acf44b7d'

{
	"didGameStart": true,
	"currentPack": -1,
	"players": [
		{
			"playerName": "bot",
			"seatNumber": 0
		},
		{
			"playerName": "dr4fter",
			"id": "1jce8jtwyfu",
			"isReadyToStart": true,
			"seatNumber": 1
		},
		{
			"playerName": "bot",
			"seatNumber": 2
		},
		{
			"playerName": "bot",
			"seatNumber": 3
		},
		{
			"playerName": "bot",
			"seatNumber": 4
		},
		{
			"playerName": "bot",
			"seatNumber": 5
		},
		{
			"playerName": "bot",
			"seatNumber": 6
		},
		{
			"playerName": "bot",
			"seatNumber": 7
		}
	]
}
```

**Comment by HerveH44** [comment]
Hi @dev-id ,

I just pushed a fix, the problem was that the bots didn't have any card. 

Thanks for your feedback! 

Hervé
